mail/cache.go

```diff
@@ -92,108 +108,164 @@ func (a *App) CacheMessages() error {
 	if err != nil {
 		return err
 	}
 
-	// We want messages in this mailbox with UIDs we haven't already cached.
-	criteria := &imap.SearchCriteria{}
-	if len(cached) > 0 {
-		uidset := imap.UIDSet{}
-		for _, cd := range cached {
-			for _, label := range cd.Labels {
-				if label.Name == mailbox {
-					uidset.AddNum(label.UID)
-				}
+	alreadyHaveSlice := []imap.UID{}
+	alreadyHaveMap := map[imap.UID]bool{}
+	for _, cm := range cached {
+		for _, label := range cm.Labels {
+			if label.Name == mailbox {
+				alreadyHaveSlice = append(alreadyHaveSlice, label.UID)
+				alreadyHaveMap[label.UID] = true
 			}
 		}
-
-		nc := imap.SearchCriteria{}
-		nc.UID = []imap.UIDSet{uidset}
-		criteria.Not = append(criteria.Not, nc)
 	}
 
-	// these are the UIDs of the new messages
-	uidsData, err := c.UIDSearch(criteria, nil).Wait()
+	// We sync the flags of all UIDs we have already cached.
+	err = a.syncFlags(alreadyHaveSlice, c, email, mailbox, cached)
 	if err != nil {
-		return fmt.Errorf("searching for uids: %w", err)
+		return err
 	}
 
-	uids := uidsData.AllUIDs()
-	if len(uids) == 0 {
-		a.AsyncLock()
-		a.Update()
-		a.AsyncUnlock()
-		return nil
+	// We filter out the UIDs that are already cached.
+	uids = slices.DeleteFunc(uids, func(uid imap.UID) bool {
+		return alreadyHaveMap[uid]
+	})
+	// We only cache in baches of 100 UIDs per mailbox to allow us to
+	// get to multiple mailboxes quickly. We want the last UIDs since
+	// those are typically the most recent messages.
+	if len(uids) > 100 {
+		uids = uids[len(uids)-100:]
 	}
-	return a.CacheUIDs(uids, c, email, mailbox, dir, cached, cacheFile)
+
+	return a.cacheUIDs(uids, c, email, mailbox, dir, cached)
+}
+
+// cleanCache removes cached messages from the given mailbox if
+// they are not part of the given list of UIDs.
+func (a *App) cleanCache(cached map[string]*CacheMessage, email string, mailbox string, uids []imap.UID) error {
+	for id, cm := range cached {
+		cm.Labels = slices.DeleteFunc(cm.Labels, func(label Label) bool {
+			return label.Name == mailbox && !slices.Contains(uids, label.UID)
+		})
+		// We can remove the message since it is removed from all mailboxes.
+		if len(cm.Labels) == 0 {
+			delete(cached, id)
+		}
+	}
+	a.imapMu[email].Lock()
+	err := a.saveCacheFile(cached, email)
+	a.imapMu[email].Unlock()
+	return err
 }
 
-// CacheUIDs caches the messages with the given UIDs in the context of the
+// cacheUIDs caches the messages with the given UIDs in the context of the
 // other given values, using an iterative batched approach that fetches the
 // five next most recent messages at a time, allowing for concurrent mail
 // modifiation operations and correct ordering.
-func (a *App) CacheUIDs(uids []imap.UID, c *imapclient.Client, email string, mailbox string, dir string, cached map[string]*CacheData, cacheFile string) error {
+func (a *App) cacheUIDs(uids []imap.UID, c *imapclient.Client, email string, mailbox string, dir string, cached map[string]*CacheMessage) error {
```

Comment:
this is a deeply nested long function -- consider useful subroutines or other logical structure.